### PR TITLE
cpeditor: init at 6.10.1

### PR DESCRIPTION
**Pull request by wineee**
###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/editors/cpeditor/default.nix [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  src = fetchFromGitHub {
+    owner = "cpeditor";
+    repo = "cpeditor";
+    rev = version;
+    sha256 = "sha256-yfHiTslKPRmFkSIj1WvaD4ZruL1azqjNAYbag1k2RzU=";
+    fetchSubmodules = true;
+  };
+
+  nativeBuildInputs = [ cmake ninja pkg-config qt5.wrapQtAppsHook ];
+  buildInputs = [ qt5.qtbase qt5.qttools python3 ];
```
Comment (bobby285271) [inline pkgs/applications/editors/cpeditor/default.nix:24]: I guess `python3` should go to `nativeBuildInputs`.

File: pkgs/applications/editors/cpeditor/default.nix [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+  preConfigure = ''
```
Comment (bobby285271) [inline pkgs/applications/editors/cpeditor/default.nix:26]: Maybe `postPatch`?

File: pkgs/applications/editors/cpeditor/default.nix [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    sed 's,/bin/bash,${stdenv.shell},' -i src/Core/Runner.cpp
```
Comment (bobby285271) [inline pkgs/applications/editors/cpeditor/default.nix:27]: Maybe use `substituteInPlace`?

File: pkgs/applications/editors/cpeditor/default.nix [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  '';
+
+  meta = with lib; {
+    description = "An IDE specially designed for competitive programming";
+    homepage = "https://cpeditor.org";
+    license = licenses.gpl3Plus;
```
Comment (bobby285271) [inline pkgs/applications/editors/cpeditor/default.nix:33]: Missing `maintainers`.

File: pkgs/applications/editors/cpeditor/default.nix (hunk 3 above)
Comment (legendofmiracles) [reply pkgs/applications/editors/cpeditor/default.nix:27]: Should also be `runtimeShell` instead of stdenv.shell

**Comment by bobby285271** [comment]
Failed to build on x86_64-darwin, shall we set platforms to linux only?

```
FAILED: CMakeFiles/diff_match_patch.dir/third_party/diff_match_patch/diff_match_patch.cpp.o
/nix/store/vn148wjd5f92yjbmlbz0rbi195daqc8a-clang-wrapper-7.1.0/bin/clang++ -DQT_CORE_LIB -DQT_NO_DEBUG -I/tmp/nix-build-cpeditor-6.9.3.drv-0/source/build -I/tmp/nix-build-cpeditor-6.9.3.drv-0/source -I/tmp/nix-build-cpeditor-6.9.3.drv-0/source/build/diff_match_patch_autogen/include -I/tmp/nix-build-cpeditor-6.9.3.drv-0/source/generated -I/tmp/nix-build-cpeditor-6.9.3.drv-0/source/src -isystem /nix/store/8mkvcnwm55ahy2lra5krhc1zxbicx6wx-qtbase-5.12.10-dev/include/QtCore -isystem /nix/store/8mkvcnwm55ahy2lra5krhc1zxbicx6wx-qtbase-5.12.10-dev/./mkspecs/macx-clang -O3 -DNDEBUG -mmacosx-version-min=10.14 -fPIC -std=gnu++14 -MD -MT CMakeFiles/diff_match_patch.dir/third_party/diff_match_patch/diff_match_patch.cpp.o -MF CMakeFiles/diff_match_patch.dir/third_party/diff_match_patch/diff_match_patch.cpp.o.d -o CMakeFiles/diff_match_patch.dir/third_party/diff_match_patch/diff_match_patch.cpp.o -c /tmp/nix-build-cpeditor-6.9.3.drv-0/source/third_party/diff_match_patch/diff_match_patch.cpp
/tmp/nix-build-cpeditor-6.9.3.drv-0/source/third_party/diff_match_patch/diff_match_patch.cpp:2038:47: error: no member named 'SkipEmptyParts' in namespace 'Qt'
  QStringList text = textline.split("\n", Qt::SkipEmptyParts);
                                          ~~~~^
1 error generated.
```

**Comment by wineee** [comment]
I'm sure cpeditor can be built on Mac，It seems to be a problem with the QT version.
This enum(Qt::SkipEmptyParts) was introduced in Qt 5.14, and it shoud be "QString::SkipEmptyParts" in old version.
https://doc.qt.io/qt-5/qt.html#SplitBehaviorFlags-enum

I don't use darwin, can't guarantee to solve the problem

**Comment by wineee** [comment]
QProcess::splitCommand was introduced in Qt 5.15

**Comment by wineee** [comment]
>[120/120] Linking CXX executable cpeditor.app/Contents/MacOS/cpeditorYJK5W5UP7/qrc_macos-resources.cpp.o
installing
install flags: install
ninja: error: unknown target 'install'
builder for '/nix/store/c9xcfvsm3hghwbky4mvkk9fsmndbddns-cpeditor-6.10.1.drv' failed with exit code 1
error: build of '/nix/store/c9xcfvsm3hghwbky4mvkk9fsmndbddns-cpeditor-6.10.1.drv' failed

I don't know how to solve it，I don't use MacOS. 
Set platforms to linux only now

**Comment by bobby285271** [comment]
Do we still need to pin Qt version?

**Comment by wineee** [comment]
> Do we still need to pin Qt version?

cpeditor exactly need Qt 5.15 or higher (Qt 6), I think it's better to pin the Qt version.
https://cpeditor.org/docs/installation/build-from-source/

File: pkgs/top-level/all-packages.nix [hunk 5]
```diff
@@ -24505,6 +24505,8 @@ with pkgs;
     ffmpeg = ffmpeg-full;
   };
 
+  cpeditor = callPackage ../applications/editors/cpeditor { };
```
Comment (bobby285271) [inline pkgs/top-level/all-packages.nix:24508]:
```suggestion
  cpeditor = libsForQt515.callPackage ../applications/editors/cpeditor { };
```

File: pkgs/applications/editors/cpeditor/default.nix [hunk 6]
```diff
@@ -0,0 +1,38 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, pkg-config
+, qt515
```
Comment (bobby285271) [inline pkgs/applications/editors/cpeditor/default.nix:5]:
```suggestion
, qtbase
, qttools
, wrapQtAppsHook
```

File: pkgs/applications/editors/cpeditor/default.nix [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  src = fetchFromGitHub {
+    owner = "cpeditor";
+    repo = "cpeditor";
+    rev = version;
+    sha256 = "sha256-SIREoOapaZTLtqi0Z07lKmNqF9a9qIpgGxuhqaY3yfU=";
+    fetchSubmodules = true;
+  };
+
+  nativeBuildInputs = [ cmake ninja pkg-config qt515.wrapQtAppsHook python3 ];
+  buildInputs = [ qt515.qtbase qt515.qttools ];
```
Comment (bobby285271) [inline pkgs/applications/editors/cpeditor/default.nix:25]:
```suggestion
  nativeBuildInputs = [ cmake ninja pkg-config wrapQtAppsHook python3 ];
  buildInputs = [ qtbase qttools ];
```

**Review by bobby285271 — APPROVED** [review]
Basic functionality looks good to me.

**Pull request merged**